### DISCUSSION TITLE
[@types/node] `utils.parseArgs` should use optional field mark `?` instead of `undefined`

### DISCUSSION THREAD

**Discussion by Conaclos** [DefinitelyTyped/DefinitelyTyped#70766] [Issues with a @types package] [answered]
Given the following code:

```ts
import * as util from "node:util";

const { values, positionals } = util.parseArgs({
  options: {
    help: { short: "h", type: "boolean", default: false },
    version: { type: "boolean" },
  }
});
```

The type of `value` is:

```
{
  help: boolean,
  version: number | undefined
}
```

If we use [TypeScript's `exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes), this is an incorrect type.
The type should be:

```
{
  help: boolean,
  version?: number
}
```

This will be expanded to the following type when `exactOptionalPropertyTypes` is turned off:

```
{
  help: boolean,
  version?: number | undefined
}
```

I discoevred the issue because I enabled `exactOptionalPropertyTypes` and noticed that the following code was failing:

```ts
import * as util from "node:util";

type Options: {
  help: boolean,
  version: number,
};

const { values, positionals } = util.parseArgs({
  options: {
    help: { short: "h", type: "boolean", default: false },
    version: { type: "boolean" },
  }
});
// Invalid assignment because `number | undefined` is not assignable to `number`.
const options: Partial<Options> = value;
```

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
@bakkot:

- do you have any thoughts on [this approach](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/77620203ba2e1b2c79e115a968e625c2446a606b#diff-6df94e39183e1ff02508a9c8876392358c90e8a634776ff280ab4a074c378550)?
- looks like this behaviour hasn't changed in the Node API since defaults were first added in v18.11.0 – was there a specific indication for only committing #70420 to the v22 branch?

**Reply by bakkot** [reply]
> * do you have any thoughts on [this approach](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/77620203ba2e1b2c79e115a968e625c2446a606b#diff-6df94e39183e1ff02508a9c8876392358c90e8a634776ff280ab4a074c378550)?

If it gives the right results I'm all for it! I notice there's still `| undefined` on the RHS, which it would be nice to get rid of if possible but I don't know why that's happening in the first place.

It could definitely use parentheses after the RHS of the `LongOption in keyof O as ...`.

> * was there a specific indication for only committing [[node:util]: for `parseArgs`, setting a default makes the value defined #70420](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70420) to the v22 branch?

I just didn't notice that there were multiple copies, my bad.

**Reply by Renegade334** [reply]
> I notice there's still `| undefined` on the RHS, which it would be nice to get rid of if possible but I don't know why that's happening in the first place.

That's just an artifact of `exactOptionalPropertyTypes` being `false` on the tests' tsconfig. If a user has it as `true`, those disappear.

**Comment by Renegade334** [comment, marked as answer]
Merged in #70781.
